externs/shaka/player.js

```diff
@@ -1334,6 +1345,23 @@ shaka.extern.ManifestConfiguration;
  * @property {number} liveSyncPanicThreshold
  *   Number of seconds that playback stays in panic mode after a rebuffering.
  *   Defaults to <code>60</code>
+ * @property {boolean} liveSyncDynamicTargetLatency
+ *   If <code>true</code>, dynamic latency for live sync is enabled. When
+ *   enabled, the target latency will be adjusted closer to the min latency
+ *   when playback is stable
+ *   (see <code>liveSyncDynamicTargetLatencyStabilityThreshold</code>). If
+ *   there are rebuffering events, then the target latency will move towards
+ *   the max latency value in increments of
+ *   <code>liveSyncDynamicTargetLatencyRebufferIncrement</code>. Defaults to
+ *   <code>false</code>.
+ * @property {number} liveSyncDynamicTargetLatencyRebufferIncrement
+ *   The value, in seconds, to increment the target latency towards
+ *   <code>liveSyncMaxLatency</code> after a rebuffering event. Defaults to
+ *   <code>0.5</code>.
+ * @property {number} liveSyncDynamicTargetLatencyStabilityThreshold
+ *   Number of seconds after a rebuffering before we are considered stable and
```

Comment:
I think you have a good overview of the changes. I've also previousl written a description of this here https://github.com/shaka-project/shaka-player/issues/6131#issuecomment-1906934434.

For increment to go towards min latency is half of the distance between target latency and min latency.
Additonally, it isn't just backing off in 0.5s increments, but rather in a multiple of 0.5s per each rebuffering event. So, if there's a bunch of rebuffering, we'll increment further back. https://github.com/shaka-project/shaka-player/pull/6193/files#diff-7abf4aa2639707ce10a19703663841fae6c5638809faf4dfb2f1894791f154ddR5694.

I think that the pendulum that you mention is definitely a possibility, though, I assume that folks that may be using this option will probably want to tweak their target and minimum settings away from the defaults. It probably doesn't make much difference with the default min, max, and target latency values.
Also, I'm not sure that a pendulum here is necessarily an issue. During playback we do want to react to occasions when there are rebuffering, so, back off on being close to the live edge and if playback is stable, we can move closer to the edge. Given the unpredicatiblity of the network, I think it's reasonable, particularly for an opt-in feature. Some internal tests, a feature like this decreased the live latency and reduced the rebuffering count.

Reply:
maybe if the rebuffer count is high we should stick to the max latency and stop trying to move it towards min?

Reply:
>  I'm not sure that a pendulum here is necessarily an issue. During playback we do want to react to occasions when there are rebuffering, so, back off on being close to the live edge and if playback is stable, we can move closer to the edge. Given the unpredicatiblity of the network, I think it's reasonable, particularly for an opt-in feature. Some internal tests, a feature like this decreased the live latency and reduced the rebuffering count.

I hear you.  But I wonder what the behavior of this will be further from the median experience.  We may not know that until it is deployed and instrumented across a large number of sessions.  Is your "internal testing" based on real user populations in an A/B experiment?  Or just "seems good" from your fast connection at work or home?  I mean no offense in this.  I'm always forced to rely on the latter myself, but I just think we should be clear about the limitations of it when we discuss our risk/reward assessments.

> I assume that folks that may be using this option will probably want to tweak their target and minimum settings away from the defaults. It probably doesn't make much difference with the default min, max, and target latency values.

While I'm sure plenty of developers are power users who understand the nuances and can tweak all these things, I bet there are still many who don't have a deep enough knowledge or experience to do so.  Or who would Deploy an updated player without noticing that this feature is in it.  Are there better defaults we could pick?  I feel like we should assume that even if it's opt-in right now, it could become opt-out some day, and we should strive for defaults that are a good starting point for most scenarios, with low risk if possible.

> maybe if the rebuffer count is high we should stick to the max latency and stop trying to move it towards min?

Maybe so.  Maybe something like this, though I'm not certain what.  I feel strongly that something should exist to stop a rebuffering cycle if it occurs.

I feel the same about latency as resolution: better to be a little father from the bleeding edge if you can avoid buffering in most cases.

But reasonable people could certainly disagree with me on that.  I also don't have any user session data to back up my philosophies on any of this.  So I would defer to real data if anyone has some.

Reply:
I totally see your concerns as well. Definitely a tricky situation trying to cater to power users while not letting others accidentally ruin their user experience.

I believe the tests were done with a network emulator and a bunch of network traces, but getting confirmation. Either way, it's definitely not the same as a full A/B experiment, though, definitely better than someone just hitting play on their laptop a bunch of times.

Would `liveSyncDynamicTargetLatencyRebufferThreshold`, where once the rebuffering count cross this boundary, it will choose the max latency and stick there without trying to push closer to the edge again assuage your concerns?

------------

Also, given that this change is essentially two in one: switch to target latency, and dynamic target latency. With shaka v5 looming, if we don't have agreement on the dynamic part, would just the target latency be able to land? Since dynamic target latency could later be added as a feature, if it doesn't happen before v5 is done. Happy to separate this PR into two if that makes it more likely for the breaking change to land in time for v5.

Reply:
Confirmed my interpretation was correct regarding the test runner.

Reply:
What would be the next action?

Reply:
Oops, somehow missed your message.

For next actions, there are several options. The one that would be the least amount of work is likely to add the `liveSyncDynamicTargetLatencyRebufferThreshold` or similar option. Alternatively, I can split this PR between the refactor to target latency and the new feature of target latency. This was more necessary when v5 was scheduled soon.

I'd like us to have consensus on the next course of action before I update the PR. Is `liveSyncDynamicTargetLatencyRebufferThreshold` reasonable? Perhaps with a new name. Or maybe there are other ideas for resolving @joeyparrish's concern about the pendulum. Currently, I don't have any other ideas, but I'll spend some more time thinking about it next week. I would appreciate any other suggestions.

Reply:
Another potential change is to move the dynamic target latency options into a sub config object.

Reply:
Yes, I think it's better to split it into two PRs, it would be easier to review. As for the new setup, I'll let Joey decide.